protobuf/lib/src/protobuf/generated_message.dart

```diff
@@ -300,8 +300,11 @@ abstract class GeneratedMessage {
 
   /// Clears the contents of a given field.
   ///
-  /// If it's an extension field, the Extension will be kept.
-  /// The tagNumber should be a valid tag or extension.
+  /// If it's an extension field, the value will be removed, but the Extension
+  /// will be kept.
+  ///
+  /// If the tagNumber is present as an unknown field, that field will be
+  /// removed.
```

Comment:
`GeneratedMessage.mergeFromMessage()` to call `_FieldSet._mergeFromMessage()`, which looks like this:
```dart
  void _mergeFromMessage(_FieldSet other) {
    for (FieldInfo fi in other._infosSortedByTag) {
      var value = other._values[fi.index];
      if (value != null) _mergeField(fi, value, isExtension: false);
    }
    if (other._hasExtensions) {
      var others = other._extensions;
      for (int tagNumber in others._tagNumbers) {
        var extension = others._getInfoOrNull(tagNumber);
        var value = others._getFieldOrNull(extension);
        _mergeField(extension, value, isExtension: true);
      }
    }
  
    if (other._hasUnknownFields) {
      _ensureUnknownFields().mergeFromUnknownFieldSet(other._unknownFields);
    }
  }
```

This merging does not seem to account for `oneof`s. Does it need to be updated as well? Are there more places?

Reply:
`_FieldSet._mergeFromMessage()` calls `_mergeField` that ends up calling `_setNonExtensionFieldUnchecked` that handles the oneofs

Reply:
Maybe my comment was not clear, what I meant is the following:

We seem to have an **invariant** now
  * there is at most one of the `oneof` cases is set (either a) it's set in field set b) it's set in unknown field set c) none are set)
  * whenever we add a new field value, we are guaranteed to clear the old one if one exists.


This invariant is violated because of this line:
```dart
    if (other._hasUnknownFields) {
      _ensureUnknownFields().mergeFromUnknownFieldSet(other._unknownFields);
    }
```

We might have already a oneof candidate in the fieldset (or the unknown fieldset). Yet this code blindly copies `other._unknownFields`, which possibly has a new `oneof` value, but we **do not clear** the old one.

Reply:
Ah - now I get it - thanks! I updated the merging of unknown fields to go field by field, and also update the one-ofs.

Reply:
Good catch, Martin. Thanks.